// @neo4j/graphql/dist/translate/cypher-builder/CypherBuilder.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Relationship = exports.RawParam = exports.Param = exports.NamedNode = exports.Node = exports.Call = exports.Apoc = exports.Merge = exports.Create = exports.Query = void 0;
var Query_1 = require("./statements/Query");
Object.defineProperty(exports, "Query", { enumerable: true, get: function () { return Query_1.Query; } });
var Create_1 = require("./statements/Create");
Object.defineProperty(exports, "Create", { enumerable: true, get: function () { return Create_1.Create; } });
var Merge_1 = require("./statements/Merge");
Object.defineProperty(exports, "Merge", { enumerable: true, get: function () { return Merge_1.Merge; } });
var Apoc_1 = require("./statements/Apoc");
Object.defineProperty(exports, "Apoc", { enumerable: true, get: function () { return Apoc_1.Apoc; } });
var Call_1 = require("./statements/Call");
Object.defineProperty(exports, "Call", { enumerable: true, get: function () { return Call_1.Call; } });
var Node_1 = require("./references/Node");
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return Node_1.Node; } });
Object.defineProperty(exports, "NamedNode", { enumerable: true, get: function () { return Node_1.NamedNode; } });
var Param_1 = require("./references/Param");
Object.defineProperty(exports, "Param", { enumerable: true, get: function () { return Param_1.Param; } });
Object.defineProperty(exports, "RawParam", { enumerable: true, get: function () { return Param_1.RawParam; } });
var Relationship_1 = require("./references/Relationship");
Object.defineProperty(exports, "Relationship", { enumerable: true, get: function () { return Relationship_1.Relationship; } });
