// @neo4j/graphql/dist/translate/cypher-builder/CypherASTNode.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CypherASTNode = void 0;
const CypherContext_1 = require("./CypherContext");
/** Abstract class representing a Cypher Statement in the AST */
class CypherASTNode {
    constructor(parent) {
        this.children = [];
        this.namedParams = {}; // Only for compatibility reasons
        this.parent = parent;
    }
    getRoot() {
        if (this.parent) {
            return this.parent.getRoot();
        }
        return this;
    }
    /** Visitor pattern to generate the Cypher on nested nodes */
    getCypher(context, separator = "\n") {
        Object.entries(this.namedParams).forEach(([name, param]) => {
            context.addNamedParamReference(name, param); // Only for compatibility reasons
        });
        const childrenCypher = this.children.map((value) => value.getCypher(context)).join(separator);
        return this.cypher(context, childrenCypher);
    }
    getContext(prefix) {
        return new CypherContext_1.CypherContext(prefix);
    }
    addStatement(astNode) {
        this.children.push(astNode);
        return astNode;
    }
    addNamedParams(params) {
        this.namedParams = { ...this.namedParams, ...params };
    }
    get isRoot() {
        return this.parent === undefined;
    }
}
exports.CypherASTNode = CypherASTNode;
